File: apps/web/src/domains/live/sensorValueInterpreter.js
import { SensorType, normalizeSensorType } from "./sensorTypes.js";

const fieldDefinitions = {
  accuracyMeter: { label: "위치 오차", unit: "m", order: 46 },
  altitudeMeter: { label: "고도", unit: "m", order: 43 },
  batteryPercent: { label: "배터리", unit: "%", order: 30 },
  headingDegree: { label: "방향", unit: "deg", order: 44 },
  networkState: { label: "네트워크", unit: "", order: 35 },
  speedMeterPerSecond: { label: "속도", unit: "m/s", order: 45 },
  yawDegree: { label: "Yaw", unit: "deg", order: 62 }
};

const segmentLabels = {
  angularVelocity: "각속도",
  linearAcceleration: "선가속도",
  x: "X",
  y: "Y",
  z: "Z"
};

const hiddenFieldNames = new Set([
  "alarm",
  "alarm_code",
  "alarmCode",
  "channel",
  "frameId",
  "high_alarm",
  "highAlarm",
  "latitude",
  "longitude",
  "low_alarm",
  "lowAlarm",
  "positionAvailable",
  "scale_code",
  "scaleCode",
  "valid"
]);

const sensorTypeOrders = {
  [SensorType.GAS]: 10,
  [SensorType.BATTERY]: 30,
  network: 35,
  [SensorType.POSITION]: 40,
  [SensorType.IMU]: 50,
  [SensorType.ODOMETRY]: 60,
  [SensorType.POINT_CLOUD]: 70,
  spatial: 75,
  [SensorType.UNKNOWN]: 90
};

const gasLabelOrders = {
  CO: 0.01,
  H2S: 0.02,
  O2: 0.03,
  CH4: 0.04,
  TEMP: 0.05,
  HUM: 0.06
};

export function interpretSensorSampleValue(descriptor, sample) {
  if (!sample) {
    return [];
  }
  const normalizedDescriptor = normalizeDescriptor(descriptor, sample);
  if (normalizedDescriptor.sensorType === SensorType.GAS) {
    return interpretGasSampleValue(normalizedDescriptor, sample);
  }
  return interpretDefaultSampleValue(normalizedDescriptor, sample);
}

function interpretDefaultSampleValue(descriptor, sample) {
  const value = getSampleValue(sample);
  const receivedAt = sample.timestamp ?? sample.receivedAt;
  if (value && typeof value === "object" && !Array.isArray(value)) {
    const readings = [];
    appendObjectReadings(readings, descriptor, value, "", receivedAt);
    return readings;
  }

  return [createReading(descriptor, {
    fieldPath: descriptor.sensorId,
    receivedAt,
    value
  })];
}

function interpretGasSampleValue(descriptor, sample) {
  const value = getSampleValue(sample);
  const receivedAt = sample.timestamp ?? sample.receivedAt;
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return interpretDefaultSampleValue(descriptor, sample);
  }
  const reading = numberOrNull(value.concentration);
  if (reading === null) {
    return interpretDefaultSampleValue(descriptor, sample);
  }
  return [createReading(descriptor, {
    fieldPath: "concentration",
    label: descriptor.label,
    receivedAt,
    unit: stringOrEmpty(value.unit) || descriptor.unit,
    value: reading
  })];
}

function appendObjectReadings(readings, descriptor, value, basePath, receivedAt) {
  Object.entries(value).forEach(([fieldName, fieldValue]) => {
    const fieldPath = basePath ? `${basePath}.${fieldName}` : fieldName;
    if (hiddenFieldNames.has(fieldName)) {
      return;
    }
    if (fieldValue && typeof fieldValue === "object" && !Array.isArray(fieldValue)) {
      appendObjectReadings(readings, descriptor, fieldValue, fieldPath, receivedAt);
      return;
    }
    readings.push(createReading(descriptor, {
      fieldPath,
      receivedAt,
      value: fieldValue
    }));
  });
}

function createReading(descriptor, { alarmLevel = "normal", fieldPath, label, receivedAt, unit, value }) {
  const fieldName = getLastPathSegment(fieldPath);
  const fieldDefinition = fieldDefinitions[fieldName];
  return {
    alarmLevel,
    fieldPath,
    key: `${descriptor.sensorId}:${fieldPath}`,
    label: label ?? makeReadingLabel(descriptor, fieldPath, fieldDefinition),
    order: metricOrder(descriptor, fieldDefinition),
    receivedAt,
    unit: unit ?? fieldDefinition?.unit ?? descriptor.unit ?? "",
    value
  };
}

function normalizeDescriptor(descriptor, sample) {
  const sensorId = descriptor?.sensorId ?? sample?.sensorId ?? "sensor.unknown";
  return {
    label: descriptor?.label ?? sensorId,
    sensorId,
    sensorType: normalizeSensorType(descriptor?.sensorType ?? inferSensorType(sensorId)),
    unit: descriptor?.unit ?? ""
  };
}

function metricOrder(descriptor, fieldDefinition) {
  if (descriptor.sensorType === SensorType.GAS) {
    return sensorTypeOrders[SensorType.GAS] + gasLabelOrder(descriptor.label);
  }
  return fieldDefinition?.order ?? (sensorTypeOrders[descriptor.sensorType] ?? sensorTypeOrders.unknown);
}

function gasLabelOrder(label) {
  return gasLabelOrders[normalizeLabelKey(label)] ?? 0.9;
}

function normalizeLabelKey(label) {
  return String(label ?? "").trim().toUpperCase();
}

function getSampleValue(sample) {
  if (!sample) {
    return null;
  }
  if (sample.values !== null && sample.values !== undefined) {
    return sample.values;
  }
  if (sample.objectKey) {
    return sample.objectKey;
  }
  return null;
}

function makeReadingLabel(descriptor, fieldPath, fieldDefinition) {
  if (fieldDefinition) {
    return fieldDefinition.label;
  }
  const segments = fieldPath.split(".");
  if (segments.length >= 2) {
    const readableSegments = segments.map((segment) => segmentLabels[segment] ?? segment);
    return `${descriptor.label} ${readableSegments.join(" ")}`;
  }
  if (fieldPath === descriptor.sensorId) {
    return descriptor.label;
  }
  return `${descriptor.label} ${segmentLabels[fieldPath] ?? fieldPath}`;
}

function getLastPathSegment(fieldPath) {
  return fieldPath.split(".").at(-1) ?? fieldPath;
}

function numberOrNull(value) {
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}

function stringOrEmpty(value) {
  return typeof value === "string" ? value.trim() : "";
}

function inferSensorType(sensorId) {
  const normalizedSensorId = String(sensorId).toLowerCase();
  if (normalizedSensorId.includes("position")) {
    return SensorType.POSITION;
  }
  if (normalizedSensorId.includes("imu")) {
    return SensorType.IMU;
  }
  if (normalizedSensorId.includes("odometry")) {
    return SensorType.ODOMETRY;
  }
  if (normalizedSensorId.includes("battery")) {
    return SensorType.BATTERY;
  }
  if (normalizedSensorId.includes("network")) {
    return "network";
  }
  if (normalizedSensorId.includes("gas")) {
    return SensorType.GAS;
  }
  return SensorType.UNKNOWN;
}

File: apps/server/internal/domain/sensor_value_interpreter.go
package domain

import (
	"encoding/json"
	"sort"
	"strings"
)

type SensorValueReading struct {
	FieldPath string  `json:"fieldPath"`
	Label     string  `json:"label"`
	Order     float64 `json:"order"`
	Unit      string  `json:"unit,omitempty"`
	Value     any     `json:"value"`
}

type sensorValueFieldDefinition struct {
	Label string
	Order float64
	Unit  string
}

var hiddenSensorValueFields = map[string]struct{}{
	"alarm":             {},
	"alarm_code":        {},
	"alarmCode":         {},
	"channel":           {},
	"frameId":           {},
	"high_alarm":        {},
	"highAlarm":         {},
	"latitude":          {},
	"longitude":         {},
	"low_alarm":         {},
	"lowAlarm":          {},
	"positionAvailable": {},
	"scale_code":        {},
	"scaleCode":         {},
	"valid":             {},
}

var sensorValueFieldDefinitions = map[string]sensorValueFieldDefinition{
	"accuracyMeter":       {Label: "위치 오차", Order: 46, Unit: "m"},
	"altitudeMeter":       {Label: "고도", Order: 43, Unit: "m"},
	"batteryPercent":      {Label: "배터리", Order: 30, Unit: "%"},
	"headingDegree":       {Label: "방향", Order: 44, Unit: "deg"},
	"networkState":        {Label: "네트워크", Order: 35},
	"speedMeterPerSecond": {Label: "속도", Order: 45, Unit: "m/s"},
	"yawDegree":           {Label: "Yaw", Order: 62, Unit: "deg"},
}

var sensorValueSegmentLabels = map[string]string{
	"angularVelocity":    "각속도",
	"linearAcceleration": "선가속도",
	"x":                  "X",
	"y":                  "Y",
	"z":                  "Z",
}

var sensorTypeReadingOrder = map[SensorType]float64{
	SensorTypeGas:        10,
	SensorTypeBattery:    30,
	SensorTypePosition:   40,
	SensorTypeIMU:        50,
	SensorTypeOdometry:   60,
	SensorTypePointCloud: 70,
	SensorTypeUnknown:    90,
}

var gasReadingLabelOrder = map[string]float64{
	"CO":   0.01,
	"H2S":  0.02,
	"O2":   0.03,
	"CH4":  0.04,
	"TEMP": 0.05,
	"HUM":  0.06,
}

func InterpretSensorSampleValue(descriptor SensorDescriptor, sample SensorSample) []SensorValueReading {
	value := decodeSensorSampleValue(sample.Values)
	if value == nil && sample.ObjectKey != "" {
		value = sample.ObjectKey
	}
	if value == nil {
		return []SensorValueReading{}
	}
	sensorType, _ := ParseSensorType(descriptor.SensorType)
	if sensorType == SensorTypeGas {
		return sortedSensorValueReadings(interpretGasSensorValue(descriptor, sample, value))
	}
	return sortedSensorValueReadings(interpretDefaultSensorValue(descriptor, sample, value))
}

func interpretGasSensorValue(descriptor SensorDescriptor, sample SensorSample, value any) []SensorValueReading {
	objectValue, ok := value.(map[string]any)
	if !ok {
		return interpretDefaultSensorValue(descriptor, sample, value)
	}
	concentration, ok := finiteNumber(objectValue["concentration"])
	if !ok {
		return interpretDefaultSensorValue(descriptor, sample, value)
	}
	unit := descriptor.Unit
	if valueUnit, ok := objectValue["unit"].(string); ok && strings.TrimSpace(valueUnit) != "" {
		unit = strings.TrimSpace(valueUnit)
	}
	return []SensorValueReading{{
		FieldPath: "concentration",
		Label:     nonEmptyString(descriptor.Label, descriptor.SensorID),
		Order:     sensorValueReadingOrder(descriptor, "concentration"),
		Unit:      unit,
		Value:     concentration,
	}}
}

func interpretDefaultSensorValue(descriptor SensorDescriptor, _ SensorSample, value any) []SensorValueReading {
	if objectValue, ok := value.(map[string]any); ok {
		readings := []SensorValueReading{}
		appendSensorObjectReadings(&readings, descriptor, "", objectValue)
		return readings
	}
	return []SensorValueReading{{
		FieldPath: descriptor.SensorID,
		Label:     nonEmptyString(descriptor.Label, descriptor.SensorID),
		Order:     sensorValueReadingOrder(descriptor, descriptor.SensorID),
		Unit:      descriptor.Unit,
		Value:     value,
	}}
}

func appendSensorObjectReadings(readings *[]SensorValueReading, descriptor SensorDescriptor, basePath string, objectValue map[string]any) {
	for fieldName, fieldValue := range objectValue {
		if _, hidden := hiddenSensorValueFields[fieldName]; hidden {
			continue
		}
		fieldPath := fieldName
		if basePath != "" {
			fieldPath = basePath + "." + fieldName
		}
		if nestedObject, ok := fieldValue.(map[string]any); ok {
			appendSensorObjectReadings(readings, descriptor, fieldPath, nestedObject)
			continue
		}
		if fieldValue == nil || fieldValue == "" {
			continue
		}
		*readings = append(*readings, SensorValueReading{
			FieldPath: fieldPath,
			Label:     makeSensorValueLabel(descriptor, fieldPath),
			Order:     sensorValueReadingOrder(descriptor, fieldPath),
			Unit:      sensorValueUnit(descriptor, fieldPath),
			Value:     fieldValue,
		})
	}
}

func makeSensorValueLabel(descriptor SensorDescriptor, fieldPath string) string {
	fieldName := lastPathSegment(fieldPath)
	if definition, ok := sensorValueFieldDefinitions[fieldName]; ok {
		return definition.Label
	}
	if fieldPath == descriptor.SensorID {
		return nonEmptyString(descriptor.Label, descriptor.SensorID)
	}
	segments := strings.Split(fieldPath, ".")
	for index, segment := range segments {
		if label, ok := sensorValueSegmentLabels[segment]; ok {
			segments[index] = label
		}
	}
	return strings.TrimSpace(nonEmptyString(descriptor.Label, descriptor.SensorID) + " " + strings.Join(segments, " "))
}

func sensorValueUnit(descriptor SensorDescriptor, fieldPath string) string {
	fieldName := lastPathSegment(fieldPath)
	if definition, ok := sensorValueFieldDefinitions[fieldName]; ok && strings.TrimSpace(definition.Unit) != "" {
		return definition.Unit
	}
	return descriptor.Unit
}

func sensorValueReadingOrder(descriptor SensorDescriptor, fieldPath string) float64 {
	fieldName := lastPathSegment(fieldPath)
	if definition, ok := sensorValueFieldDefinitions[fieldName]; ok {
		return definition.Order
	}
	sensorType, _ := ParseSensorType(descriptor.SensorType)
	order := sensorTypeReadingOrder[sensorType]
	if order == 0 {
		order = sensorTypeReadingOrder[SensorTypeUnknown]
	}
	if sensorType == SensorTypeGas {
		order += gasReadingLabelOrder[strings.ToUpper(strings.TrimSpace(descriptor.Label))]
	}
	return order
}

func sortedSensorValueReadings(readings []SensorValueReading) []SensorValueReading {
	sort.SliceStable(readings, func(leftIndex, rightIndex int) bool {
		left := readings[leftIndex]
		right := readings[rightIndex]
		if left.Order != right.Order {
			return left.Order < right.Order
		}
		if left.Label != right.Label {
			return left.Label < right.Label
		}
		return left.FieldPath < right.FieldPath
	})
	return readings
}

func decodeSensorSampleValue(rawValue json.RawMessage) any {
	if len(rawValue) == 0 {
		return nil
	}
	var value any
	if err := json.Unmarshal(rawValue, &value); err != nil {
		return nil
	}
	return value
}

func finiteNumber(value any) (float64, bool) {
	switch number := value.(type) {
	case float64:
		return number, true
	case float32:
		return float64(number), true
	case int:
		return float64(number), true
	case int64:
		return float64(number), true
	default:
		return 0, false
	}
}

func lastPathSegment(value string) string {
	segments := strings.Split(value, ".")
	return segments[len(segments)-1]
}

func nonEmptyString(value string, fallback string) string {
	value = strings.TrimSpace(value)
	if value == "" {
		return strings.TrimSpace(fallback)
	}
	return value
}
